hypothesis-python/src/hypothesis/internal/conjecture/dfa/lstar.py

```diff
@@ -219,3 +250,58 @@ def transition(self, i, c):
             self.__rows_to_states[row] = result
         self.__transition_cache[key] = result
         return result
+
+
+class IntegerNormalizer:
+    """A class for replacing non-negative integers with a
+    "canonical" value that is equivalent for all relevant
+    purposes."""
+
+    def __init__(self):
+        # We store canonical values as a sorted list of integers
+        # with each value being treated as equivalent to the largest
+        # integer in the list that is below it.
+        self.__values = IntList([0])
+
+    def __repr__(self):
+        return "IntegerNormalizer(%r)" % (list(self.__values),)
+
+    def __copy__(self):
+        result = IntegerNormalizer()
+        result.__values = IntList(self.__values)
+        return result
+
+    def normalize(self, value):
+        """Return the canonical integer considered equivalent
+        to ``value``."""
+        i = bisect_right(self.__values, value) - 1
+        assert i >= 0
+        return self.__values[i]
+
+    def distinguish(self, value, test):
+        """Checks whether ``test`` gives the same answer for
+        ``value`` and ``self.normalize(value)``. If it does
+        not, updates the list of canonical values so that
+        it does.
+
+        Returns True if and only if this makes a change to
+        the underlying canonical values."""
+        canonical = self.normalize(value)
+        value_test = test(value)
+        if canonical == value or test(canonical) == value_test:
```

Comment:
Is there any reason to call `test` before doing the `canonical == value` test here?

Reply:
No, I suppose not. I wanted to avoid calling it multiple times, but it makes sense to put the `canonical == value` check before that rather than having it combined in the `or`.